Correct louvain formula

z3y50n opened the pull request:
Closes #5175 

After playing around with the modularity formula I figured we weren't accounting for the modularity cost when we remove a node from its original community.

The following example can prove the correctness of the new formula:

Initially we have a partition and we move node 2 from the 0th community into the 1st community. We use NetworkX's `modularity` function to measure the actual gain.

```Python
>>> import networkx as nx
>>> from networkx.algorithms.community import louvain_communities, quality
>>> G = nx.karate_club_graph()
>>> partition = louvain_communities(G, weight="weight", seed=123)
>>> print(partition)
[{1, 2, 3, 7, 12, 13}, {0, 4, 5, 6, 10, 11, 16, 17, 19, 21}, {23, 24, 25, 27, 28, 31}, {32, 33, 8, 9, 14, 15, 18, 20, 22, 26, 29, 30}]
>>> before = quality.modularity(G, partition, weight="weight")
>>> partition[0].remove(2)
>>> partition[1].add(2)
>>> print(partition)
[{1, 3, 7, 12, 13}, {0, 2, 4, 5, 6, 10, 11, 16, 17, 19, 21}, {23, 24, 25, 27, 28, 31}, {32, 33, 8, 9, 14, 15, 18, 20, 22, 26, 29, 30}]
>>> after = quality.modularity(G, partition, weight="weight")
>>> print(after - before)
-0.0553494124922696
```
The below snippet computes both the remove and insert parameters to calculate the gain.
``` Python3
Stot = 0
RStot = 0
for n in partition[1]:
    if n == 2:
        continue
    Stot += G.degree(n, weight="weight")
for n in partition[0]:
    RStot += G.degree(n, weight="weight")

r_wt = 0
wt = 0
for u, w in G[2].items():
    if u in partition[0]:
        r_wt += w["weight"]
    if u in partition[1]:
        wt += w["weight"]

m = G.size(weight="weight")
ki = G.degree(2, weight="weight")
```

Finally:
```Python3
>>> gain = (wt) / (m) - (Stot * ki) / (2 * m ** 2) # Original formula didn't take into consideration the removal of node from initial community
>>> remove_cost = -r_wt / m + (RStot * ki) / (2 * m ** 2)
>>> print(gain) # Incorrect Result
-0.011131725417439706
>>> print(remove_cost)
-0.04421768707482993
>>> print(remove_cost + gain) # Actual modularity gain from whole movement process
-0.05534941249226964
```

PS. I have added a test for the issue's case which doesn't loop forever with the new formula. Feel free to propose any better test cases that I might miss.

rossbar reviewed (commented):
Thanks @z3y50n , this definitely looks like it's on the right track, though I haven't looked at the formula updates in detail.

The first thing I tried was the case reported in #5704 and it's still hanging. It could be that it's due to a completely separate issue, in which case I'd vote to reopen #5704 and evaluate it separately; however, if the expectation was that this PR would also fix that issue then I think we should add it to the tests here and figure out why it's still not working.

z3y50n commented:
I tried a similar approach for the directed case ([here](https://gist.github.com/z3y50n/386e4487e6831f39d9dfeb9dc6a4e40f) is a notebook showing my work) and figured the correct formula which matches the one in the directed paper.

Unfortunately #5704 is still halting and as of now I am not sure of the reason (could it be a defect of the algorithm itself or another bug in the code?).

rossbar commented:
Thanks @z3y50n - at this point I think we should split the directed/undirected cases into two separate PRs, as it seems like the fixups for the undirected case are ready. I'd propose the following:
 1. Undo the changes related to the directed case from this PR
 2. At that point, this one will be ready for review, and we can work on getting it in!
 3. Reopen #5704 with a note that the issue appears to be separate (enough) from #5175

Re: a strategy for figuring out why the directed case is still getting stuck, it might be worth zooming out from the modularity update computation and just stepping through the algorithm applied to the case in #5704 with a debugger. This process can be a bit tedious, but at the very least it should help identify where exactly the "infinite loop" manifests. That's just one idea though, a first-principles analysis is obviously preferable but if you've hit a roadblock there a little interactive debugging might be informative!

z3y50n commented:
So after trial and error I figured the last missing piece of the puzzle. When calculating the first term in the formula I was considering only the out degree of the candidate node to nodes inside the community, but we need to consider both the in and out degrees instead. I have updated the [gist](https://gist.github.com/z3y50n/386e4487e6831f39d9dfeb9dc6a4e40f) and also added test cases for the issues that were causing the infinite loops.

rossbar commented on networkx/algorithms/community/louvain.py:
```diff
@@ -219,10 +219,18 @@ def _one_level(G, m, partition, resolution=1, is_directed=False, seed=None):
+        nbrs = {}
+        for u in G:
+            nbrs[u] = defaultdict(float)
+            for _, n, wt in G.out_edges(u, data=True):
+                nbrs[u][n] += wt["weight"]
+            for n, _, wt in G.in_edges(u, data=True):
+                nbrs[u][n] += wt["weight"]
```
Just a nit, but you can pass the edge attr name directly to the `data` kwarg to make this a little cleaner:

```suggestion
            for _, n, wt in G.out_edges(u, data="weight"):
                nbrs[u][n] += wt
            for n, _, wt in G.in_edges(u, data="weight"):
                nbrs[u][n] += wt
```

rossbar commented on networkx/algorithms/community/tests/test_louvain.py:
```diff
@@ -43,6 +43,58 @@ def test_partition():
     assert part == partition
 
 
+def test_directed_partition():
```
This should probably be parametrized or split up into two separate tests, but that's a minor organizational nit and not a blocker!

z3y50n replied:
I actually wanted to parametrize this function but the edgelists to construct the graphs are too big and I wasn't sure how it would look in terms of styling. Maybe one better way is to create a class?

Also, if we split  the test cases (which I am in favor of) do you think naming the functions something like `test_5175_issue` be good?

rossbar replied:
> but the edgelists to construct the graphs are too big and I wasn't sure how it would look in terms of styling.

I agree - black will likely force this test to be less readable, so I'd say let's not worry about it for now :). There are ways to "trick" black into looking better here (i.e. define the edgelists externally and pass them into `parametrize`), but I'd say that's better for a follow-up PR - the current tests LGTM (minus the missing `G.add_nodes_from`).

rossbar commented on networkx/algorithms/community/tests/test_louvain.py:
```diff
@@ -43,6 +43,58 @@ def test_partition():
+    from issues #5175 and #5704
+    """
+    G = nx.DiGraph()
+    H = nx.DiGraph()
```
The original failure reported in #5175 was also sensitive to node insertion order, at least by my testing (see [this comment](https://github.com/networkx/networkx/issues/5175#issuecomment-965602312)) so I think we also need:

```suggestion
    H = nx.DiGraph()
    G.add_nodes_from(range(10))
```

rossbar replied:
I'm going to go ahead and put this one in, just to verify that the tests still pass afterwards!

rossbar reviewed (commented):
Thanks for this @z3y50n , very nice work chasing this down! I had one nit-picky suggestion for the code that you can safely ignore :). On the testing front, I think there is one piece missing to reproduce the initial reported failing case (IIRC it was sensitive to node order) so I think that should be added just to be extra safe - other than that LGTM!

rossbar commented:
Ah ok - so changing the node ordering (by adding the `add_nodes_from`) for the first test example *does* avoid the infinite loop as hoped, but also results in different communities. Is this expected @z3y50n ? If so, we can simply change the `expected` values for these tests and call it done :)

rossbar reviewed (approved):
Okay, I think that's the last of the test updates, LGTM. Thanks a ton for chasing this down @z3y50n!

FWIW I also verified (locally) that this fixes #5858 - that particular case is quite large (100's of nodes and 1000's of edges) so I don't think it needs to be added as another test case.
